src/ck-core/cklocation.C

```diff
@@ -3026,8 +3039,17 @@ void CkLocMgr::immigrate(CkArrayElementMigrateMessage *msg)
 	//Create a record for this element
 	CkLocRec *rec=createLocal(idx,true,msg->ignoreArrival,false /* home told on departure */ );
 	
+	envelope *env = UsrToEnv(msg);
+	CmiAssert(CkpvAccess(newZCPupGets).empty()); // Ensure that vector is empty
 	//Create the new elements as we unpack the message
 	pupElementsFor(p,rec,CkElementCreation_migrate);
+	if(!CkpvAccess(newZCPupGets).empty()) {
+		// newZCPupGets is not empty, rgets need to be launched
+		// newZCPupGets is populated with NcpyOperationInfo during pupElementsFor by pup_buffer calls that require Rgets
+		// Issue Rgets using the populated newZCPupGets vector
+		zcPupIssueRgets(msg->id, this);
```

Comment:
What happens if the rgets don't complete inline (eg the migration is across the network)? Where do we handle the case where completion is asynchronous?

Reply:
I think this is the issue Evan is seeing

Reply:
Right, for asynchronous Gets (Rgets, which don't complete inline), in Charm++, I just buffer the messages to the target chares to not allow those messages to execute since the data is not ready. It's only after the completion of the Rgets in `zcPupGetCompleted` that the buffered messages are executed because the data is ready. 

However, I see that for `isomalloc`, it looks like you'd want a capability where the pup (with `isUnpacking` instance) function is not called until the data transfer is complete. Currently, the `pup` function is called during unpacking (just as before) because we'd want the input parameter `buffer`(the pointer) that the user wants the destination region to be assigned to.

Reply:
The `isUnpacking` PUP::er would still need to be called before the data transfer so that the receiver knows what the proper `allocate` function is. The problem is that execution of the TCharm chare resumes before the data transfer completes.

Reply:
Okay, so, you're saying an entry method of TCharm resumes before the data transfer completes?